executor/adapter.go

```diff
@@ -305,6 +310,24 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 		logutil.Logger(ctx).Error("execute sql panic", zap.String("sql", a.GetTextToLog()), zap.Stack("stack"))
 	}()
 
+	failpoint.Inject("assertStaleTSO", func(val failpoint.Value) {
+		if n, ok := val.(int64); ok {
+			startTS := oracle.ExtractPhysical(a.TSO)
+			if n != startTS {
+				panic("different tso")
+			}
+			failpoint.Return()
+		}
+	})
+	failpoint.Inject("assertStaleTSOWithTolerance", func(val failpoint.Value) {
+		if n, ok := val.(int64); ok {
+			startTS := oracle.ExtractPhysical(a.TSO)
+			if startTS <= n-2 || n+2 <= startTS {
```

Comment:
I think the timestamp that `oracle.ExtractPhysical` returns is in millisecond, and the tolerance is 2 seconds if I remember correctly.

Reply:
Great catch.